File: src/main/java/facts/exp/BinOp.java
package facts.exp;

import facts.Fact;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;

/**
 * Binary operation fact
 */
public class BinOp extends Expr {

    /** Operator */
    private String operator;
    /** Expression 1 */
    private Expr firstExp;
    /** Expression 2 */
    private Expr secondExp;

    public BinOp(String operator, Expr firstExp, Expr secondExp) {
        this.operator = operator;
        this.firstExp = firstExp;
        this.secondExp = secondExp;
    }

    public String getOperator() {
        return operator;
    }

    public void setOperator(String operator) {
        this.operator = operator;
    }

    public Expr getFirstExp() {
        return firstExp;
    }

    public void setFirstExp(Expr firstExp) {
        this.firstExp = firstExp;
    }

    public Expr getSecondExp() {
        return secondExp;
    }

    public void setSecondExp(Expr secondExp) {
        this.secondExp = secondExp;
    }

    public List<Fact> toFactList() {
        List<Fact> factList = new ArrayList<>();
        factList.addAll(firstExp.toFactList());
        factList.addAll(secondExp.toFactList());
        factList.add(this);
        return factList;
    }

    @Override
    public String toString() {
        return String.format("(%s) %s (%s)", firstExp, operator, secondExp);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        BinOp binOp = (BinOp) o;
        return Objects.equals(operator, binOp.operator) && Objects.equals(firstExp, binOp.firstExp) && Objects.equals(secondExp, binOp.secondExp);
    }

    @Override
    public int hashCode() {
        return Objects.hash(operator, firstExp, secondExp);
    }

    @Override
    public List<Expr> getChildren() {
        return Arrays.asList(firstExp, secondExp);
    }

    @Override
    public void replace(Expr oldExp, Expr newExp) {
        if (firstExp.equals(oldExp)) {
            firstExp = newExp;
        }
        if (secondExp.equals(oldExp)) {
            secondExp = newExp;
        }
    }
}

File: src/main/java/facts/exp/BinOp.scala
package facts.exp

import jdk.javadoc.doclet.DocletEnvironment.ModuleMode
import java.util
import java.util.Objects

/** Binary operation fact
  */
case class BinOp(
    operator: BinOperator.Value,
    var firstExp: Expr,
    var secondExp: Expr
) extends Expr {
  def this(operatorStr: String, firstExp: Expr, secondExp: Expr) = this(BinOperator.fromBil(operatorStr), firstExp, secondExp)
  override def toString = String.format("(%s) %s (%s)", firstExp, operator, secondExp)
  override def getChildren = util.Arrays.asList(firstExp, secondExp)

  // TODO update so the member vars can be vals
  override def replace(oldExp: Expr, newExp: Expr) = {
    if (firstExp == oldExp) firstExp = newExp
    if (secondExp == oldExp) secondExp = newExp
  }

  override def vars = firstExp.vars ++ secondExp.vars
}

object BinOperator extends Enumeration {
  type Operator = Value
  // arithmetic operators
  val Addition: Operator = Value("+")
  val Subtraction: Operator = Value("-")
  val Multiplication: Operator = Value("*")
  val Division: Operator = Value("/")
  val Modulo: Operator = Value("%")
  val BitwiseAnd: Operator = Value("&") // todo
  val BitwiseOr: Operator = Value("|") // todo
  val BitwiseXor: Operator = Value("xor") // todo
  val LogicalShiftLeft: Operator = Value("<<") // todo
  val LogicalShiftRight: Operator = Value(">>") // todo
  val ArithmeticShiftRight: Operator = Value(">>>") // todo
  // logical operators
  val Equality: Operator = Value("==")
  val NonEquality: Operator = Value("!=")
  val LessThan: Operator = Value("<")
  val LessThanOrEqual: Operator = Value("<=")

  def fromBil(bilStr: String): Value = bilStr match {
    // note at the moment we do not distinguish between signed and unsigned
    // arithmetic operators
    case "+" => Addition
    case "-" => Subtraction
    case "*" => Multiplication
    case "/" => Division
    case "%" => Modulo
    case "&" => BitwiseAnd
    case "|" => BitwiseOr
    case "xor" => BitwiseXor
    case "<<" => LogicalShiftLeft
    case ">>" => LogicalShiftRight
    case ">>>" => ArithmeticShiftRight
    // logical operators
    case "=" => Equality
    case "<>" => NonEquality
    case "<" => LessThan
    case "<=" => LessThanOrEqual
  }
}
